src/lib/expression/expression_utils.hpp

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <vector>
+#include <unordered_map>
+#include <queue>
+
+#include "abstract_expression.hpp"
+#include "parameter_expression.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+class LQPColumnExpression;
+class TransactionContext;
+
+bool expressions_equal(const std::vector<std::shared_ptr<AbstractExpression>>& expressions_a,
+                             const std::vector<std::shared_ptr<AbstractExpression>>& expressions_b);
+
+/**
+ * Utility to compare vectors of Expressions from different LQPs
+ */
+bool expressions_equal_to_expressions_in_different_lqp(
+const std::vector<std::shared_ptr<AbstractExpression>> &expressions_left,
+const std::vector<std::shared_ptr<AbstractExpression>> &expressions_right,
+const LQPNodeMapping& node_mapping);
+
+/**
+ * Utility to compare two Expressions from different LQPs
+ */
+bool expression_equal_to_expression_in_different_lqp(const AbstractExpression& expression_left,
+                       const AbstractExpression& expression_right,
+                       const LQPNodeMapping& node_mapping);
+
+
+std::vector<std::shared_ptr<AbstractExpression>> expressions_copy(
+  const std::vector<std::shared_ptr<AbstractExpression>>& expressions);
+
+std::vector<std::shared_ptr<AbstractExpression>> expressions_copy_and_adapt_to_different_lqp(
+  const std::vector<std::shared_ptr<AbstractExpression>>& expressions,
+  const LQPNodeMapping& node_mapping);
+
+std::shared_ptr<AbstractExpression> expression_copy_and_adapt_to_different_lqp(
+  const AbstractExpression& expression,
+  const LQPNodeMapping& node_mapping);
+
+/**
+ * Makes all ColumnExpressions points to their equivalent in a copied LQP
+ */
+void expression_adapt_to_different_lqp(
+  std::shared_ptr<AbstractExpression>& expression,
+  const LQPNodeMapping& node_mapping);
+
+std::shared_ptr<LQPColumnExpression> expression_adapt_to_different_lqp(
+  const LQPColumnExpression& lqp_column_expression,
+  const LQPNodeMapping& node_mapping);
+
+std::string expression_column_names(const std::vector<std::shared_ptr<AbstractExpression>> &expressions);
+
+/**
+ * @tparam Expression   Either `std::shared_ptr<AbstractExpression>` or `const std::shared_ptr<AbstractExpression>`
+ * @tparam Visitor      Functor called with every sub expression as a param.
+ *                      Return true to traverse to sub expressions of the current expression
```

Comment:
I really like this function. Just as a little comment: If you see the code of the visitor function somewhere else without knowing this comment. It would be more intuitive (at least for me) to traverse the subexpressions if you return false. Mostly because the functions are often named like find_xy, that's why it is at first a little bit irritating to see a return false when you found xy.